### PR TITLE
Warning nodes

### PR DESCRIPTION
**Pull request by SHKnudsen**
### Purpose

Adding Warning class to Application category + nodes and tests

### Reviewers

@radumg 

#### CanGetFailingElementsFromWarning
CanGetFailingElementsFromWarning

#### CanGetSeverityOfWarning
CanGetSeverityOfWarning

#### CanGetWarningDescription
CanGetWarningDescription

#### CanGetWarningsFromCurrentDocument
CanGetWarningsFromCurrentDocument

#### CanGetWarningsFromWarningType
CanGetWarningsFromWarningType

### CODE REVIEW

File: src/Libraries/RevitNodes/Application/Warning.cs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+namespace Revit.Application
+{
+    public class Warning
+    {
+        /// <summary>
+        /// Internal reference to the Document
+        /// </summary>
+        internal Autodesk.Revit.DB.FailureMessage InternalWarning { get; private set; }
+
+        internal Warning(Autodesk.Revit.DB.FailureMessage warning)
```
Comment (radumg) [inline src/Libraries/RevitNodes/Application/Warning.cs]: i'd rename the `warning` parameter to `failure`

File: src/Libraries/RevitNodes/Application/Warning.cs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        {
+            InternalWarning = warning;
+        }
+
+        [IsVisibleInDynamoLibrary(false)]
+        public static List<Warning> GetWarningByGuid(string guid)
```
Comment (radumg) [inline src/Libraries/RevitNodes/Application/Warning.cs:25]: if we pass in a null or empty string, do we get the expected error ? if yes, ok, if not, we need to add a `string.IsNullOrWhiteSpace` check

File: src/Libraries/RevitNodes/Application/Warning.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        {
+            var warnings = DocumentManager.Instance.CurrentDBDocument
+                .GetWarnings()
+                .GroupBy(warn => warn.GetFailureDefinitionId().Guid.ToString())
+                .ToList();
+
+            var keys = warnings.Select(key => key.First().GetFailureDefinitionId().Guid.ToString());
+            var dict = keys.Zip(warnings, (k, v) => new { Key = k, Value = v })
+                .ToDictionary(x => x.Key, x => x.Value);
```
Comment (radumg) [inline src/Libraries/RevitNodes/Application/Warning.cs]: not sure we need to do this again - feels like there should be a `ToDictionary` or `ToKeyValuePairs` LINQ method we can use above, this is essentially doing additional work.

Comment (radumg) [reply src/Libraries/RevitNodes/Application/Warning.cs]: actually, if all you need is to retrieve warning by Guid, why are we grouping in the first place ? we should be able to select the one(s) we need with a `Select` or `Where(x=>string.Equals(x.GetFailureDefinitionId().Guid.ToString(), guid))`

File: src/Libraries/RevitNodesUI/RevitDropDown.cs [hunk 4]
```diff
@@ -929,5 +930,74 @@ public override IEnumerable<AssociativeNode> BuildOutputAst(List<AssociativeNode
...
     }
 
-        
+    [NodeName("All Warnings of Type")]
+    [NodeCategory(BuiltinNodeCategories.REVIT_WARNING)]
+    [NodeDescription("AllWarningsOfTypeDescription", typeof(DSRevitNodesUI.Properties.Resources))]
+    [IsDesignScriptCompatible]
+    public class AllWarningsOfType : RevitDropDownBase
+    {
+        private const string outputName = "Warning";
```
Comment (radumg) [inline src/Libraries/RevitNodesUI/RevitDropDown.cs]: if we're outputting multiple warnings, this should be plural

File: src/Libraries/RevitNodesUI/RevitDropDown.cs [hunk 5]
```diff
@@ -929,5 +930,74 @@ public override IEnumerable<AssociativeNode> BuildOutputAst(List<AssociativeNode
...
+        protected override SelectionState PopulateItemsCore(string currentSelection)
+        {
+            Items.Clear();
+            Autodesk.Revit.DB.Document document = DocumentManager.Instance.CurrentDBDocument;
+
+            //find all unique warnings in the project
+            var warnings = DocumentManager.Instance.CurrentDBDocument
+                .GetWarnings()
+                .GroupBy(warn => warn.GetFailureDefinitionId().Guid.ToString())
+                .Select(group => group.First())
```
Comment (radumg) [inline src/Libraries/RevitNodesUI/RevitDropDown.cs:957]: why are we only getting the first group ?

File: src/Libraries/RevitNodesUI/RevitDropDown.cs [hunk 6]
```diff
@@ -929,5 +930,74 @@ public override IEnumerable<AssociativeNode> BuildOutputAst(List<AssociativeNode
...
+
+        [JsonConstructor]
+        public AllWarningsOfType(IEnumerable<PortModel> inPorts, IEnumerable<PortModel> outPorts) : base(outputName, inPorts, outPorts) { }
+
+        protected override SelectionState PopulateItemsCore(string currentSelection)
+        {
+            Items.Clear();
+            Autodesk.Revit.DB.Document document = DocumentManager.Instance.CurrentDBDocument;
+
+            //find all unique warnings in the project
```
Comment (radumg) [inline src/Libraries/RevitNodesUI/RevitDropDown.cs]: code style issue : add a space after `//`

File: src/Libraries/RevitNodesUI/RevitDropDown.cs [hunk 7]
```diff
@@ -929,5 +930,74 @@ public override IEnumerable<AssociativeNode> BuildOutputAst(List<AssociativeNode
...
+            var warnings = DocumentManager.Instance.CurrentDBDocument
+                .GetWarnings()
+                .GroupBy(warn => warn.GetFailureDefinitionId().Guid.ToString())
+                .Select(group => group.First())
+                .ToList();
+
+            if (warnings.Count < 1)
+                throw new InvalidOperationException("No warnings in the current document");
```
Comment (radumg) [inline src/Libraries/RevitNodesUI/RevitDropDown.cs]: this text should be in `.resx`

File: src/Libraries/RevitNodesUI/RevitDropDown.cs [hunk 8]
```diff
@@ -929,5 +930,74 @@ public override IEnumerable<AssociativeNode> BuildOutputAst(List<AssociativeNode
...
+
+            for (int i = 0; i < warnings.Count; i++)
+            {
+                var warningText = TruncateString(warnings[i].GetDescriptionText(), 50);
```
Comment (radumg) [inline src/Libraries/RevitNodesUI/RevitDropDown.cs]: value `50` should be a constant that has a descriptive name or comment above saying what/why it's used for.
I can venture a guess why, but would be good to be told exactly why

File: src/Libraries/RevitNodesUI/RevitDropDown.cs (hunk 5 above)
Comment (SHKnudsen) [reply src/Libraries/RevitNodesUI/RevitDropDown.cs:957]: grouping all warnings by their guid, then getting the first warning of each group so i get a list of unique warnings to display in the dropdown

Comment (radumg) [reply src/Libraries/RevitNodesUI/RevitDropDown.cs:957]: you can use the LINQ `ToDistinct()` for this, should be better/faster than grouping then selecting again i think

**Pull request merged**